File: Demo/newobject/GameSdk.cpp
#include <jni.h>
#include <string>
#include <vector>

#include "rapidjson/document.h"
#include "rapidjson/prettywriter.h"

#include "JniHelper.h"
#include "GameSdk.h"

#define JAVA_CLASS "com/seastar/spg/Interface/GameSDK"


void Java_com_seastar_spg_Interface_GameSDK_onFbInitCb(JNIEnv *env, jclass obj, jint flags)
{
	if (flags)
	{
		// init success proc
	}
	else
	{
		// init failure proc
	}
}

void Java_com_seastar_spg_Interface_GameSDK_onGoogleInitCb(JNIEnv *env, jclass obj, jint flags)
{
	if (flags)
	{
		// init success proc
	}
	else
	{
		// init failure proc
	}
}

void Java_com_seastar_spg_Interface_GameSDK_onFbLoginCb(JNIEnv *env, jclass obj, jint flags, jstring user)
{
	if (flags)
	{
		std::string strUser("");
		const char *pszUser = env->GetStringUTFChars(user, NULL);
		strUser = pszUser;
		env->ReleaseStringUTFChars(user, pszUser);

		rapidjson::Document doc;
		doc.Parse(strUser.c_str());

		std::string id = doc["id"].GetString();
		std::string name = doc["name"].GetString();
		std::string picture = doc["picture"].GetString();
		std::string email = doc["email"].GetString();
		std::string first_name = doc["first_name"].GetString();
		std::string last_name = doc["last_name"].GetString();
		std::string middle_name = doc["middle_name"].GetString();
		std::string gender = doc["gender"].GetString();
		std::string location = doc["location"].GetString();

		// do login success proc
	}
	else
	{
		// do login failure proc
	}

}

void Java_com_seastar_spg_Interface_GameSDK_onGoogleLoginCb(JNIEnv *env, jclass obj, jint flags, jstring user)
{
	bool isLogined = false;
	std::string strUser("");

	if (flags == 1)
	{
		const char *pszUser = env->GetStringUTFChars(user, NULL);
		strUser = pszUser;
		env->ReleaseStringUTFChars(user, pszUser);

		rapidjson::Document doc;
		doc.Parse(strUser.c_str());

		std::string id = doc["id"].GetString();
		std::string ageRange = doc["ageRange"].GetString();
		std::string birthday = doc["birthday"].GetString();
		std::string currentLocation = doc["currentLocation"].GetString();
		std::string displayName = doc["displayName"].GetString();
		std::string image = doc["image"].GetString();
		std::string language = doc["language"].GetString();
		std::string name = doc["name"].GetString();
		std::string nickName = doc["nickName"].GetString();
		std::string url = doc["url"].GetString();

		// do login success proc
	}
	else
	{
		// do login failure proc
	}
}

void Java_com_seastar_spg_Interface_GameSDK_onGooglePayCb(JNIEnv *env, jclass obj, jint flags, jstring sku,
		jstring itemType, jstring googleOrder, jstring signature, jstring purchaseOriginalData)
{
	if (flags)
	{
		std::string strSku("");
		std::string strItemType("");
		std::string strGoogleOrder("");
		std::string strSignature("");
		std::string strPurchaseOriginalData("");

		const char *pszSku = env->GetStringUTFChars(sku, NULL);
		const char *pszItemType = env->GetStringUTFChars(itemType, NULL);
		const char *pszGoogleOrder = env->GetStringUTFChars(googleOrder, NULL);
		const char *pszSignature = env->GetStringUTFChars(signature, NULL);
		const char *pszPurchaseOriginalData = env->GetStringUTFChars(purchaseOriginalData, NULL);

		strSku = pszSku;
		strItemType = pszItemType;
		strGoogleOrder = pszGoogleOrder;
		strSignature = pszSignature;
		strPurchaseOriginalData = pszPurchaseOriginalData;

		env->ReleaseStringUTFChars(sku, pszSku);
		env->ReleaseStringUTFChars(itemType, pszItemType);
		env->ReleaseStringUTFChars(googleOrder, pszGoogleOrder);
		env->ReleaseStringUTFChars(signature, pszSignature);
		env->ReleaseStringUTFChars(purchaseOriginalData, pszPurchaseOriginalData);

		// pay success proc
	}
	else
	{
		// pay failure proc
	}
}

void Java_com_seastar_spg_Interface_GameSDK_onGoogleConsumeCb(JNIEnv *env, jclass obj, jint flags, jstring sku)
{
	if (flags)
	{
		std::string strSku("");
		const char *pszSku = env->GetStringUTFChars(sku, NULL);

		strSku = pszSku;

		env->ReleaseStringUTFChars(sku, pszSku);

		// consume success proc
	}
	else
	{
		// consume failure proc
	}

}

void Java_com_seastar_spg_Interface_GameSDK_onQueryGoogleInventory(JNIEnv *env, jclass obj,
		jint flags, jstring skus)
{
	if (flags)
	{
		std::string strSkus("");
		const char *pszSkus = env->GetStringUTFChars(skus, NULL);

		strSkus = pszSkus;

		env->ReleaseStringUTFChars(skus, pszSkus);

		rapidjson::Document doc;
		doc.Parse(strSkus.c_str());

		const rapidjson::Value &value = doc["arr"];
		for (int i = 0; i < value.Size(); i++)
		{
			std::string description = value[i]["description"].GetString();
			std::string price = value[i]["price"].GetString();
			std::string sku = value[i]["sku"].GetString();
			std::string title = value[i]["title"].GetString();
			std::string type = value[i]["type"].GetString();
			std::string priceAmountMicros = value[i]["priceAmountMicros"].GetString();
			std::string priceCurrencyCode = value[i]["priceCurrencyCode"].GetString();
			std::string googleOrder = value[i]["googleOrder"].GetString();
			std::string signature = value[i]["signature"].GetString();
			std::string purchaseOriginalData = value[i]["purchaseOriginalData"].GetString();

			// 每个循环一个购买的商品信息
		}

		

		// query success proc
	}
	else
	{
		// query success failure proc
	}
}

void doFbInit()
{
	JniMethodInfo t;
	if (JniHelper::getStaticMethodInfo(t, JAVA_CLASS, "doFbInit", "()V"))
	{
		t.env->CallStaticVoidMethod(t.classID, t.methodID);
		t.env->DeleteLocalRef(t.classID);
	}
}

void doGoogleInit()
{
	JniMethodInfo t;
	if (JniHelper::getStaticMethodInfo(t, JAVA_CLASS, "doGoogleInit", "()V"))
	{
		t.env->CallStaticVoidMethod(t.classID, t.methodID);
		t.env->DeleteLocalRef(t.classID);
	}
}

void doFbLogin()
{
	JniMethodInfo t;
	if (JniHelper::getStaticMethodInfo(t, JAVA_CLASS, "doFbLogin", "()V"))
	{
		t.env->CallStaticVoidMethod(t.classID, t.methodID);
		t.env->DeleteLocalRef(t.classID);
	}
}

void doGoogleLogin()
{
	JniMethodInfo t;
	if (JniHelper::getStaticMethodInfo(t, JAVA_CLASS, "doGoogleLogin", "()V"))
	{
		t.env->CallStaticVoidMethod(t.classID, t.methodID);
		t.env->DeleteLocalRef(t.classID);
	}
}

void doFbLogout()
{
	JniMethodInfo t;
	if (JniHelper::getStaticMethodInfo(t, JAVA_CLASS, "doFbLogout", "()V"))
	{
		t.env->CallStaticVoidMethod(t.classID, t.methodID);
		t.env->DeleteLocalRef(t.classID);
	}
}

void doGoogleLogout()
{
	JniMethodInfo t;
	if (JniHelper::getStaticMethodInfo(t, JAVA_CLASS, "doGoogleLogout", "()V"))
	{
		t.env->CallStaticVoidMethod(t.classID, t.methodID);
		t.env->DeleteLocalRef(t.classID);
	}
}

void doGooglePay(std::string &sku)
{
	JniMethodInfo t;
	if (JniHelper::getStaticMethodInfo(t, JAVA_CLASS, "doGooglePay",
			"(Ljava/lang/String;)V"))
	{
		jstring jSku = t.env->NewStringUTF(sku.c_str());

		t.env->CallStaticVoidMethod(t.classID, t.methodID, jSku);

		t.env->DeleteLocalRef(jSku);

		t.env->DeleteLocalRef(t.classID);
	}
}

void doGoogleConsume(std::string &sku, std::string &itemType, std::string &signature, std::string &purchaseOriginalData)
{
	JniMethodInfo t;
	if (JniHelper::getStaticMethodInfo(t, JAVA_CLASS, "doGoogleConsume",
			"(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;)V"))
	{
		jstring jSku = t.env->NewStringUTF(sku.c_str());
		jstring jItemType = t.env->NewStringUTF(itemType.c_str());
		jstring jSignature = t.env->NewStringUTF(signature.c_str());
		jstring jPurchaseOriginalData = t.env->NewStringUTF(purchaseOriginalData.c_str());

		t.env->CallStaticVoidMethod(t.classID, t.methodID, jSku, jItemType, jSignature, jPurchaseOriginalData);

		t.env->DeleteLocalRef(jSku);
		t.env->DeleteLocalRef(jItemType);
		t.env->DeleteLocalRef(jSignature);
		t.env->DeleteLocalRef(jPurchaseOriginalData);

		t.env->DeleteLocalRef(t.classID);
	}
}

void queryGoogleInventory(std::vector<std::string> &skus)
{
	if (skus.size() == 0)
		return;

	std::string strSku;

	for (int i = 0; i < skus.size(); i++)
	{
		strSku = strSku + " " + skus[i];
	}
	strSku = strSku.substr(1, strSku.length());

	JniMethodInfo t;
	if (JniHelper::getStaticMethodInfo(t, JAVA_CLASS, "queryGoogleInventory",
			"(Ljava/lang/String;)V"))
	{
		jstring jSku = t.env->NewStringUTF(strSku.c_str());

		t.env->CallStaticVoidMethod(t.classID, t.methodID, jSku);

		t.env->DeleteLocalRef(jSku);

		t.env->DeleteLocalRef(t.classID);
	}
}

File: Demo/newobject/seastar-spg-sdk/src/com/seastar/spg/Interface/GameSdk.java
package com.seastar.spg.Interface;

import java.util.ArrayList;
import java.util.List;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import com.seastar.spg.sdk.*;
import com.seastar.spg.sdk.Facebook.FacebookUserModel;
import com.seastar.spg.sdk.Facebook.OnInitCallBack;
import com.seastar.spg.sdk.Facebook.OnLoginCallBack;
import com.seastar.spg.sdk.Google.GoogleSkuModel;
import com.seastar.spg.sdk.Google.GoogleUserModel;

import android.app.Activity;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;

public class GameSdk {

	private static Facebook fb = new Facebook();
	private static Google google = new Google();
	private static Handler handler;

	public static void setCurActivity(Activity activity) {
		fb.setCurActivity(activity);
		google.setCurActivity(activity);
		handler = new Handler() {
			@Override
			public void handleMessage(Message msg) {
				switch (msg.what) {
				case 0:
					fb.init(new Facebook.OnInitCallBack() {

						@Override
						public void onInitCallBack(boolean isInit) {
							onFbInitCb(isInit ? 1 : 0);
						}
					});
					break;
				case 1:
					google.init(new Google.OnInitCallBack() {

						@Override
						public void onInitCallBack(boolean isInit) {
							onGoogleInitCb(isInit ? 1 : 0);
						}
					});
					break;
				case 2:
					fb.login(new Facebook.OnLoginCallBack() {

						@Override
						public void onLoginCallBack(boolean isLogin, FacebookUserModel user) {
							// TODO 自动生成的方法存根

							if (isLogin) {
								try {
									JSONObject obj = new JSONObject();
									obj.put("id", user.id);
									obj.put("name", user.name);
									obj.put("picture", user.picture);
									obj.put("email", user.email);
									obj.put("first_name", user.first_name);
									obj.put("last_name", user.last_name);
									obj.put("middle_name", user.middle_name);
									obj.put("gender", user.gender);
									obj.put("location", user.location);

									onFbLoginCb(1, obj.toString());
								} catch (JSONException e) {
									e.printStackTrace();
								}
							} else {
								onFbLoginCb(0, "");
							}
						}
					});
					break;
				case 3:
					google.login(new Google.OnLoginCallBack() {

						@Override
						public void onLoginCallBack(boolean success, GoogleUserModel googleUserModel) {
							// TODO 自动生成的方法存根
							if (success) {
								try {
									JSONObject obj = new JSONObject();
									obj.put("id", googleUserModel.id);
									obj.put("ageRange", googleUserModel.ageRange);
									obj.put("birthday", googleUserModel.birthday);
									obj.put("currentLocation", googleUserModel.currentLocation);
									obj.put("displayName", googleUserModel.displayName);
									obj.put("image", googleUserModel.image);
									obj.put("language", googleUserModel.language);
									obj.put("name", googleUserModel.name);
									obj.put("nickName", googleUserModel.nickName);
									obj.put("url", googleUserModel.url);

									onGoogleLoginCb(1, obj.toString());
								} catch (JSONException e) {
									e.printStackTrace();
								}

							} else {
								onGoogleLoginCb(0, "");
							}
						}
					});
					break;
				case 4:
					fb.logout();
					break;
				case 5:
					google.logout();
					break;
				case 6:
					google.pay(String.valueOf(msg.obj), new Google.OnPayCallBack() {

						@Override
						public void onPayCallBack(boolean success, String sku, String itemType, String googleOrder,
								String signature, String purchaseOriginalData) {
							// TODO 自动生成的方法存根
							onGooglePayCb(success ? 1 : 0, sku, itemType, googleOrder, signature, purchaseOriginalData);
						}
					});
					break;
				case 7:
					Bundle bundle = (Bundle) msg.obj;
					google.consume(bundle.getString("sku"), bundle.getString("itemtype"), bundle.getString("signature"),
							bundle.getString("purchaseOriginalData"), new Google.OnConsumeCallBack() {

						@Override
						public void onConsumeCallBack(boolean success, String sku) {
							// TODO 自动生成的方法存根
							onGoogleConsumeCb(success ? 1 : 0, sku);

						}
					});
					break;
				case 8:
					String skus = String.valueOf(msg.obj);
					
					List<String> skusList = new ArrayList<>();
					String[] skusArr = skus.split(" ");
					for (int i = 0; i < skusArr.length; i++) {
						skusList.add(skusArr[i]);
					}

					google.queryInventoryAsync(skusList, new Google.OnQueryInventoryCallBack() {

						@Override
						public void onQueryInventoryCallBack(boolean success, List<GoogleSkuModel> googleSkuModels) {
							if (success) {
								try {

									JSONObject obj = new JSONObject();
									JSONArray arr = new JSONArray();
									obj.put("skus", arr);

									for (GoogleSkuModel sku : googleSkuModels) {
										JSONObject s = new JSONObject();
										s.put("description", sku.description);
										s.put("price", sku.price);
										s.put("sku", sku.sku);
										s.put("title", sku.title);
										s.put("type", sku.type);
										s.put("priceAmountMicros", sku.priceAmountMicros);
										s.put("priceCurrencyCode", sku.priceCurrencyCode);
										s.put("googleOrder", sku.googleOrder);
										s.put("signature", sku.signature);
										s.put("purchaseOriginalData", sku.purchaseOriginalData);

										arr.put(s);
									}

									onQueryGoogleInventory(1, obj.toString());
								} catch (JSONException e) {
									e.printStackTrace();
								}

							} else {
								onQueryGoogleInventory(0, "");
							}
						}
					});
					break;
				default:
					super.handleMessage(msg);
					break;
				}
			}
		};
	}

	public static void doFbInit() {
		handler.sendEmptyMessage(0);
	}

	public static void doGoogleInit() {
		handler.sendEmptyMessage(1);
	}

	public static void doFbLogin() {
		handler.sendEmptyMessage(2);
	}

	public static void doGoogleLogin() {
		handler.sendEmptyMessage(3);
	}

	public static void doFbLogout() {
		handler.sendEmptyMessage(4);
	}

	public static void doGoogleLogout() {
		handler.sendEmptyMessage(5);
	}

	public static void doGooglePay(String sku) {
		Message msg = new Message();
		msg.what = 6;
		msg.obj = sku;
		handler.sendMessage(msg);
	}

	public static void doGoogleConsume(String sku, String itemType, String signature, String purchaseOriginalData) {
		Bundle bundle = new Bundle();
		bundle.putString("sku", sku);
		bundle.putString("itemtype", itemType);
		bundle.putString("signature", signature);
		bundle.putString("purchaseOriginalData", purchaseOriginalData);

		Message msg = new Message();
		msg.what = 7;
		msg.obj = bundle;
		handler.sendMessage(msg);
	}

	public static void queryGoogleInventory(String skus) {
		Message msg = new Message();
		msg.what = 8;
		msg.obj = skus;
		handler.sendMessage(msg);
	}

	/*
	public static void query_Skus(final ArrayList<String> skulist, OnQueryInventoryCallBack onQueryInventoryCallBack) {
		google.querySkus(skulist, new Google.OnQuerySkusDetailsCallBack() {

			String skus;

			@Override
			public void onQuerySkusDetailsCallBack(boolean success, List<GoogleSkuModel> googleSkuModel) { // TODO
																											// 自动生成的方法存根

				if (success) {
					for (int i = 0; i < skulist.size(); i++) {
						if (i == skulist.size() - 1) {
							skus = skulist.get(i);
						} else {
							skus = skulist.get(i) + " ";
						}
					}

					querySkus(skus, new Google.OnQueryInventoryCallBack() {

						@Override
						public void onQueryInventoryCallBack(boolean success, List<GoogleSkuModel> googleSkuModel) { // TODO
																														// 自动生成的方法存根

						}
					});
				} else {
					querySkus(null, null);
				}
			}

		});

	}
	*/

	public static native void onFbInitCb(int flags);

	public static native void onGoogleInitCb(int flags);

	public static native void onFbLoginCb(int flags, String fbUser);

	public static native void onGoogleLoginCb(int flags, String googleUser);

	public static native void onGooglePayCb(int flags, String sku, String itemType, String googleOrder,
			String signature, String purchaseOriginalData);

	public static native void onGoogleConsumeCb(int flags, String sku);

	public static native void onQueryGoogleInventory(int flags, String skus);

}
